### ISSUE TITLE
Binary classification is insufficient

### ISSUE THREAD

**Issue by TechnikEmpire** [TechnikEmpire/NsfwSqueezenet#6]
Given the topic of pornography, you'd think there isn't that much variation in content to warrant anything more than binary or even unary classification. However, thinking the problem over more in depth has caused me to arrive at the conclusion that multiclass is the only viable, well rounded approach.

It is necessary to develop several vategories. These could include things like penetration, male genital, female genitalia and more. Furthermore, it may be beneficial to do a SSD with this sort of multi class contextual classification.

**Comment by TechnikEmpire** [comment]
Related to #5

**Comment by jessetrana** [comment]
For what it's worth, I've been working on a similar project and have gone through perhaps similar questions.

I started with 4 categories: safe, questionable, racy, explicit. Broadly these corresponded to no questionable content, images with borderline but definitely not pornographic content (e.g. a fashion photo with a shirt open down the middle but covering breasts), images with clearly suggestive and arguably pornographic material, to images with nudity and often sexual acts.

After a while, I grew dissatisfied with the accuracy and the apparent subjectivity of some of my grading choices - particularly in say the "questionable" class. I decided to make a much more nuanced system grading on somewhat independent axes of suggestivity and exposure (as well as violence, presence of drugs, and offensive text). Grading was slower, arguably more objective. Initial checks of the grading seemed reasonable. I rated 95,000 images this way, with about 40% questionable or worse. I then bucketed the more nuanced system into a binary classification and performed training over the course of some months. Accuracy improved, but not as greatly as I would have liked. Worse, I found to my chagrin that the qualitative results when used at scale were likely poorer - more objective ratings were failing common sense checks.

So I moved back to the four class system to better capture context, but used my hard-won experience and was much more rigorous about defining what types of nudity and acts would go into which bucket (also, things like offensive text might move images to say questionable). Things are looking a bit better now, and I would actually prefer my poorer accuracy on the four classes because it often degrades more gracefully. And further, it is much, much easier to grade a single four class label than something more nuanced - and that can help reduce the toll of grading difficult images. (I can see 24 images at a shot when grading.)

At the end of the day, the thing that mattered to me with respect to classes was really what penalty I wanted to place on the classes so that I could create an appropriate weighted loss function, as well as a system that mapped well to my use case of a "kid's filter" and an "adult's filter" (whether or not questionable was included).

Hope that helps. I'm still on the journey, too. Good luck.

**Comment by TechnikEmpire** [comment]
Thanks for sharing your experience.

I've decided there's a better way to do this. I'm hesitant to divulge the whole process but basically the process involves 8K textures and blender 3d.

The solution, I believe, is less data, not more. I'm going to do multi class with a single individual per sample and absolute background subtraction. This way the network will ignore everything except what's important. Then I'll shove those samples through some sort of augmentation process to diversify the input and call it day.

While I have not completed this work, I have high hopes that this will be the answer to a proper NSFW filter that catches both plain nudity and explicit pornography.

I personally don't care about making a distinction between "adult" and "kid" filter. If its sexual at all, it must be caught, so I guess our goals diverge there.

I classify text, images, and video independently and even text is split into three subsystems for classification. I won't lean on one to infer the content of another. Its too difficult to establish a clear context.

**Comment by jessetrana** [comment]
Thanks for a preview of your plan, and best of luck!

If you're willing to share - how do you currently handle suggestive content that is not traditionally pornographic, e.g. foot fetish etc.? That type of content definitely gave me pause as I thought how to grade it.

**Comment by TechnikEmpire** [comment]
I wouldn't care about plain feet, but if theres something else in the picture (if u know what I'm saying) I would expect that object to be detected.

Again I'm thinking that the background subtraction with random augmentation should help diversify detection hopefully as extensively as people can imagine ways to pevert arbitrary things.

**Comment by TechnikEmpire** [comment]
My approach to this is like, how can I write a software that will protect the innocence of children and enable individuals from viewing nudity and pornography. If someone wants to get off looking at pics of feet or hands or something else weird that's just out of scope for me lol.

**Comment by jessetrana** [comment]
There is a new project, [NSFW JS/NSFW Model](https://github.com/GantMan/nsfw_model), that might give some inspiration. It appears he is using a 5 class system of neutral, drawings, hentai, porn, and sexy and is fine tuning both an Inception V3 and a MobileNetV2 model based on data scraper data.

**Comment by TechnikEmpire** [comment]
Thanks. No model like that will succeed. You cannot perform whole context classification. Feed the network a picture of nature where a penis is partially visible and I guarantee it classifies it as neutral. In fact feed it in a 90 degree rotated pornographic image, I bet it fails.

**Comment by xaerincl** [comment]
Have any of you tried using YOLO or another object detection algorithm instead of doing normal image classification? And train it with like dickpics or whatever you want to recognize. Im trying to solve a similar problem like you guys.

**Comment by TechnikEmpire** [comment]
Yes. Its funny you ask because right now my computer is crunching out such a network. Not using YOLO since computation is slower and maP is lower than alternatives.

**Comment by TechnikEmpire** [comment]
I'd be interested in hearing more about what your broader project is.

**Comment by xaerincl** [comment]
It is pretty similar to the NSFW project tbh. Trying to detect (hopefully in real time) "malicious content"  for now im thinking: porn, hentai, gore/violence.

tell me later about your results! and how did you get the train data for the image classification? you did it yourself?

**Comment by TechnikEmpire** [comment]
That's a trade secret unfortunately. While I published this network, the SSD network being trained now is for a commercial content filtering software.

**Comment by TechnikEmpire** [comment]
My method doesn't involve feeding the network existing, labelled pornography.

**Comment by xaerincl** [comment]
Oh damn must be a super sophisticated model then. Bad luck that you cant tell me more about it but I understand. Good luck with your model!

**Comment by jessetrana** [comment]
@xaerincl  I've thought about treating the problem as object detection, but for my use cases I encountered some things that have made me not switch yet.

1. Inference performance - My project is a Firefox plugin doing content filtering on inbound images fully client side ([Wingman Jr](https://addons.mozilla.org/en-US/firefox/addon/wingman-jr-filter/)), so the additional heft of e.g. SSD/YOLO would make an already slow model even slower.
2. Hard-to-objectify bounding boxes - As you perhaps noticed from the discussion above, @TechnikEmpire  have taken a bit different philosophical approaches to the problem with respect to what we're trying to filter. I believe certain parts of the problem would be easy to create bounding boxes for; however, for many of the things I'm trying to filter there's quite a bit of context involved and figuring out where exactly the box should be drawn or if it should be multiple boxes could be quite difficult.

If you combine these with the increased complexity in creating the dataset (along with my poor little TX1 doing the training, too for that matter), I've decided not to switch to that approach at this time.

**Comment by TechnikEmpire** [comment]
That's a good point re inference time.

However, there's such a thing as mobilenet ssd2, which is mobilenet as the base network architecture and the latter end of the overall network is ssd box detection.

Original ssd is two networks merged together. Vgg and then the invented ssd tail end. The latter does some inference on bounding boxes for the different classes you train it on, which is how its able to do multi class detection in a "single pass". Yolo works on the same principle.

Mobilenet is designed for lightweight inference (mobile) and the bounding box detection is the least process intensify aspect. Mobilenet does suffer from slightly lower accuracy that pure ssd because you have an old, tried test and heavily deep network on the front end that does exponentially more ops on many more weighted sums.

So yeah if you look at the inference time for opencv dnn with mobilenetssd you're looking at around 30 msec depending on your input size.

**Comment by TechnikEmpire** [comment]
I forgot I basically gave away my trade secret already. Find some people willing to get naked and 3D scan them and with some time and coding, you have infinite input to train your network.

You can tag vertices or place empties at target object locations and do raycasts to tell of the object is within the frame at a specific render and auto-emit the bounding boxes for the various classes. 

Plus you gave absolute perfect background subtraction so your network never learns any features that are useless. A sum of ops performed on 0 is 0 so those empty pixels do not hurt the calculation, if anything they help by becoming dead neurons so they don't tip the scales one way or another.
